Draw a random coefficient sample in MoMoInstance

andreas-schneider opened the pull request:
Currently, it is laborious to draw a random MoMoInstance random coefficient that can be used for example rendering a random model sample.

Functionality:
Draws shape, color and expression vector entries N(0,1) distributed.

This is a possible use case of the proposed functionality:
```scala
val model: MoMo = ???
val renderer = MoMoRenderer(model)
val parameter = RenderParameter.defaultSquare.withMoMo(MoMoInstance.sample(model))
val randomRendering = renderer.renderImage(parameter)
```

Commit:
Draw a random coefficient sample in MoMoInstance.

Andreas-Forster commented on src/main/scala/scalismo/faces/momo/MoMoCoefficients.scala:
```diff
@@ -17,6 +17,7 @@
 package scalismo.faces.momo
 
 import breeze.linalg.DenseVector
+import breeze.stats.distributions.Gaussian
```
If you do not move the functionality from MoMoInstance to here then this import is not needed.

Andreas-Forster commented on src/main/scala/scalismo/faces/parameters/RenderObject.scala:
```diff
@@ -141,6 +143,20 @@ object MoMoInstance {
+
+  /** create a random coefficient vector with its elements N(0,1) distributed. */
+  def sample(model: MoMo, modelURI: URI)(implicit rnd: Random): MoMoInstance = {
+    val fullShapeCoeffs = for (_ <- 0 until model.neutralModel.shape.rank) yield Gaussian(0, 1).draw()
```
Why creating a Gaussian(0,1) three times? You could also use rng.breezeRandBasis.gaussian.draw() instead.

Andreas-Forster commented on src/main/scala/scalismo/faces/parameters/RenderObject.scala:
```diff
@@ -141,6 +143,20 @@ object MoMoInstance {
   }
+
+  /** create a random coefficient vector with its elements N(0,1) distributed. */
+  def sample(model: MoMo, modelURI: URI)(implicit rnd: Random): MoMoInstance = {
```
The first part of this function should go to MoMoCoefficients.sample() and then call this function here.

Andreas-Forster commented on src/main/scala/scalismo/faces/parameters/RenderObject.scala:
```diff
@@ -141,6 +143,20 @@ object MoMoInstance {
+    val fullShapeCoeffs = for (_ <- 0 until model.neutralModel.shape.rank) yield Gaussian(0, 1).draw()
+    val fullColorCoeffs = for (_ <- 0 until model.neutralModel.color.rank) yield Gaussian(0, 1).draw()
+    val expressionRank = model.expressionModel.get.expression.rank
+    val coeff = if(expressionRank > 0) {
```
You do not need to check. The if and else branch do exactly the same for values smaller or equal zero.

Andreas-Forster reviewed (commented):
I am somehow not sure if we should provide this functionality in the MoMoInstance or MoMoCoefficients class. Of course, at the moment they have a very strict setting they are applied to in our framework but others might reuse them without the restriction of each coefficient being standard Gaussian distributed. This property is induced by using a specific model and not from the coefficients class.

BernhardEgger commented:
In this implementation the coefficients are always of full rank of the model. In practice it is often desired to have only the first x parameters. The MoMo itself has already a sample function which gives the VertexColorMesh directly. Without the possibility to use less parameters I see no benefit...

Commit:
Sample in MoMoCoefficients.

Commit:
MoMoInstance sample with specified size of vector.

andreas-schneider commented:
@Andreas-Forster 
I have incorporated your suggestions.

@BernhardEgger 
- The MoMo.sample() method can not always be used, for example when wanting to render a random sample with the MoMoRenderer. There you need to supply the coefficients.
- You are right, usefulness will be incresed when we can choose the size of the vectors. I added the functionality. You can now choose the size of MoMoCoefficients and MoMoInstance.

sschoenborn commented:
Just a side note: In `PancakeDLRGP` you have a `coefficientsDistribution` - this could be a possible concept to capture the distribution (including a `sample()` method) of the coefficients for a model

Andreas-Forster committed:
added sampleCoefficients to the trait MoMo

Andreas-Forster commented:
@sschoenborn Thank you for pointing that out. I think this is maybe the best solution for this. I provided the implementation in addition to the former one.

@BernhardEgger I think it is a rare case where you want to have coefficients with a lot of different length in one application. So given this new solution, you would truncate the model first before sampling the coefficients. This makes also the generation of instances faster afterward.

@andreas-schneider It would now be possible to write:
```scala
  val model: MoMo = ???
  val modelURI: URI = ???
  val renderer = MoMoRenderer(model)
  val parameter = RenderParameter.defaultSquare.withMoMo(MoMoInstance.fromCoefficients(model.sampleCoefficients(),modelURI))
  val randomRendering = renderer.renderImage(parameter)
```
What do you think about this solution?

(Please do not merge as we should clean one of the solutions before.)

Commit:
keep sampleCoefficients solution.

andreas-schneider commented:
Yes, the sampleCoefficients sounds good. I have updated it accordingly.

Andreas-Forster committed:
Merge branch 'master' into feature-momoinstance-random-sample
